src/Discovery/Archive/RPM.hs

```diff
@@ -14,30 +16,59 @@ import qualified Data.ByteString as BS
 import qualified Data.ByteString.Char8 as BS8
 import qualified Data.ByteString.Lazy as BL
 import qualified Data.CPIO as CPIO
+import qualified Data.Conduit.Lzma as Lzma
+import qualified Data.Conduit.Zlib as Zlib
+import Data.Foldable (asum)
 import Path
-import Prelude
 import qualified Path.IO as PIO
+import Prelude
 
 extractRpm :: Has (Lift IO) sig m => Path Abs Dir -> Path Abs File -> m ()
 extractRpm dir rpmFile = do
   res <-
-    sendIO . runResourceT . runExceptT . runConduit $
-      readRPMEntries rpmFile .| filterC (not . CPIO.isEntryDirectory) .| sinkDir
+    sendIO . runResourceT . runExceptT . runConduit $ do
+      sourceFileBS (toFilePath rpmFile) .| RPM.parseRPMC .| awaitForever (extractEntries dir)
+
   case res of
     Left parseErr -> sendIO . throwIO $ parseErr
     Right () -> pure ()
+
+-- | Extract RPM entries to a directory
+extractEntries :: (PrimMonad m, MonadThrow m, MonadIO m) => Path Abs Dir -> RPMTypes.RPM -> ConduitT i o m ()
+extractEntries dir rpm = yield rpm .| RPM.payloadC .| decompressorFor rpm .| CPIO.readCPIO .| filterC (not . CPIO.isEntryDirectory) .| sinkDir dir
```

Comment:
We're using conduit here, which allows us to do _streaming_ decompression on the RPM entries. Some notes about reading conduit code:

Whenever you see a type like `ConduitT i o m a`, the type variables mean:
- `i` - the type of elements coming from the upstream
- `o` - the type of elements we're passing downstream
- `m` - the underlying monad (ConduitT is a monad transformer)
- `a` - the arbitrary result type of this computation (like any other monad)

The way to read `.|` is to think of it as a unix pipe. Each function in between is doing a transformation on things moving down the pipe